src/modelmanager.cpp

```diff
@@ -322,23 +327,54 @@ Status ModelManager::loadModelsConfig(rapidjson::Document& configJson) {
         return StatusCode::JSON_INVALID;
     }
     std::set<std::string> modelsInConfigFile;
-    servedModelConfigs.clear();
+    std::unordered_map<std::string, ModelConfig> newModelConfigs;
     for (const auto& configs : itr->value.GetArray()) {
-        ModelConfig& modelConfig = servedModelConfigs.emplace_back();
+        ModelConfig modelConfig;
         auto status = modelConfig.parseNode(configs["config"]);
+        const auto modelName = modelConfig.getName();
         if (!status.ok()) {
-            SPDLOG_LOGGER_ERROR(modelmanager_logger, "Parsing model:{} config failed",
-                modelConfig.getName());
-            servedModelConfigs.pop_back();
+            SPDLOG_LOGGER_ERROR(modelmanager_logger, "Parsing model:{} config failed", modelName);
             continue;
         }
-        reloadModelWithVersions(modelConfig);
+        status = reloadModelWithVersions(modelConfig);
         modelsInConfigFile.emplace(modelConfig.getName());
+
+        if (status.ok()) {
+            newModelConfigs.emplace(modelName, std::move(modelConfig));
+        } else if (status == StatusCode::REQUESTED_DYNAMIC_PARAMETERS_ON_SUBSCRIBED_MODEL) {
+            SPDLOG_DEBUG("Will retry to reload model({}) after pipelines are revalidated", modelName);
+            auto it = this->servedModelConfigs.find(modelName);
+            if (it == this->servedModelConfigs.end()) {
+                continue;
+            }
+            this->gatedModelConfigs.emplace_back(std::move(modelConfig));
+            newModelConfigs.emplace(modelName, std::move(it->second));
+            this->servedModelConfigs.erase(modelName);
+        }
     }
+    this->servedModelConfigs = std::move(newModelConfigs);
     retireModelsRemovedFromConfigFile(modelsInConfigFile);
     return ovms::StatusCode::OK;
 }
 
+Status ModelManager::tryReloadGatedModelConfigs() {
+    for (auto& modelConfig : this->gatedModelConfigs) {
+        SPDLOG_DEBUG("Trying to reload model({}) configuration", modelName);
+        auto status = reloadModelWithVersions(modelConfig);
+        if (!status.ok()) {
+            continue;
+        }
+        auto it = this->servedModelConfigs.find(modelConfig.getName());
+        if (it == this->servedModelConfigs.end()) {
+            continue;
+        }
+        SPDLOG_DEBUG("Successfully retried to load new model({}) configuration after unsubscribed from pipeline", modelConfig.getName());
+        this->servedModelConfigs.at(modelConfig.getName()) = std::move(modelConfig);
+    }
+    this->gatedModelConfigs.clear();
```

Comment:
Minor:
I don't like clearing here since we strongly rely on the code as it is now. One could change it that clear is in some cases not called, or tryReloadGatedModelConfigs.

GatedModelConfigs is an information about "state of loading". Maybe it would be cleaner to move it between member functions?
It would complicate return of loadModelsConfig but I think it may be worth it.

Reply:
Done